libraries/botbuilder-dialogs-adaptive/src/adaptiveDialog.ts

```diff
@@ -200,93 +159,115 @@ export class AdaptiveDialog<O extends object = {}> extends Dialog<O> {
-                        handled = await this.evaluateRules(planning, { name: RuleDialogEventNames.activityReceived, value: undefined, bubble: false });
-                    }
-                }
-                break;
-            case RuleDialogEventNames.consultDialog:
-                // Emit event
-                handled = await this.queueFirstMatch(planning, event);
-                if (!handled) {
-                    // Dispatch activityReceived event
-                    handled = await this.evaluateRules(planning, { name: RuleDialogEventNames.activityReceived, value: undefined, bubble: false });
+                    });
+                    sequence.queueChanges(changes);
+                    handled = true;
+                } else {
+                    // Emit leading ActivityReceived event 
+                    handled = await this.processEvent(sequence, { name: AdaptiveEventNames.activityReceived, bubble: false }, true);
                 }
                 break;
-            case RuleDialogEventNames.activityReceived:
-                // Emit event
-                handled = await this.queueFirstMatch(planning, event);
-                if (!handled) {
-                    const activity = planning.context.activity;
-                    if (activity.type === ActivityTypes.Message) {
-                        // Recognize utterance
-                        const recognized = await this.onRecognize(planning.context);
-    
-                        // Dispatch utteranceRecognized event
-                        handled = await this.evaluateRules(planning, { name: RuleDialogEventNames.recognizedIntent, value: recognized, bubble: false });
-                    } else if (activity.type === ActivityTypes.Event) {
-                        // Dispatch named event that was received
-                        handled = await this.evaluateRules(planning, { name: activity.name, value: activity.value, bubble: false });
+            case AdaptiveEventNames.activityReceived:
+                const activity = sequence.context.activity;
+                if (activity.type === ActivityTypes.Message) {
+                    // Recognize utterance
+                    const recognized = await this.onRecognize(sequence.context);
+                    sequence.state.setValue('turn.recognized', recognized);
+
+                    // Emit leading RecognizedIntent event
+                    handled = await this.processEvent(sequence, { name: AdaptiveEventNames.recognizedIntent, value: recognized, bubble: false }, true);
+                } else if (activity.type === ActivityTypes.Event) {
+                    // Emit leading edge named event that was received
+                    handled = await this.processEvent(sequence, { name: activity.name, value: activity.value, bubble: false }, true);
+                } else if (activity.type === ActivityTypes.ConversationUpdate && Array.isArray(activity.membersAdded)) {
```

Comment:
I feel a bit uneasy about the adaptive dialog having specific knowledge about things like members added and such. Seems like in an ideal world, the event handling precedence along with specific handling could be in satellite classes. If this code will be somewhat stable this shouldn't be a problem though.

Reply:
I think this will eventually get externalized to another component